Remove skips from doctests (1 of 6)

zzhengnan opened the pull request:
This is one of 6 PRs that collectively addresses #7823. Please see https://github.com/dask/dask/issues/7823#issuecomment-873414398 for more context.

Links to all 6 PRs
- #7874
- #7873
- #7872
- #7865
- #7864
- #7863

jsignell commented on dask/base.py:
```diff
@@ -1225,13 +1225,13 @@ def get_collection_names(collection) -> Set[str]:
     --------
     >>> a.__dask_keys__()  # doctest: +SKIP
-    ["foo", "bar"]  # doctest: +SKIP
+    ["foo", "bar"]
```
I didn't realize that the `doctest: +SKIP` is only required on the input. This is a nice cleanup!

jsignell commented on dask/blockwise.py:
```diff
@@ -113,7 +113,7 @@ class BlockwiseDepDict(BlockwiseDep):
 
-    >>> import pandas as pd  # doctest: +SKIP
+    >>> import pandas as pd
     >>> func = lambda x: pd.read_csv(**x)  # doctest: +SKIP
```
Why not take the   `# doctest: +SKIP` off this and the class definition below as well?

jsignell commented on dask/delayed.py:
```diff
@@ -67,10 +67,10 @@ def unpack_collections(expr):
-    >>> collections  # doctest: +SKIP
-    (a, b)
+    >>> collections
+    (Delayed('a'), Delayed('b'))
```
Yay! You found a docs improvement!! :clap:

jsignell commented on dask/local.py:
```diff
@@ -61,7 +61,7 @@
 --------
 
->>> import pprint  # doctest: +SKIP
+>>> import pprint
```
I'm not sure how much value there is in unskipping an import if the whole rest of the doctest is skipped. If seems like you could probably unskip these tests though.

zzhengnan replied:
Done.

jsignell commented on dask/optimization.py:
```diff
@@ -25,8 +25,8 @@ def cull(dsk, keys):
-    >>> d = {'x': 1, 'y': (inc, 'x'), 'out': (add, 'x', 10)}    # doctest: +SKIP
-    >>> dsk, dependencies = cull(d, 'out')                      # doctest: +SKIP
+    >>> d = {'x': 1, 'y': (inc, 'x'), 'out': (add, 'x', 10)}
+    >>> dsk, dependencies = cull(d, 'out')
```
I think the following lines can be unskipped with the following:

```python
    >>> dsk  # doctest: +ELLIPSIS
    {'out': (<function add at ...>, 'x', 10), 'x': 1}
    >>> dependencies
    {'out': ['x'], 'x': []}
```

zzhengnan replied:
Great suggestion on using `+ELLIPSIS`! I'll see if I can incorporate that for the other PRs as well.

In addition to that, I see you've flipped the order of the keys in the dictionary, presumably to get the output matching. There are many instances across all 6 PRs where I wasn't able to un-skip a doctest simply because the key ordering is different even though the dict itself is identical. 

Should I simply reorder the dict in the docstring to match the test output? How can I be sure that the specific ordering I get on my machine is the "canonical" one that's going to remain unchanged going forward?

jsignell replied:
> Should I simply reorder the dict in the docstring to match the test output? How can I be sure that the specific ordering I get on my machine is the "canonical" one that's going to remain unchanged going forward?

Yeah good point. I just looked at the code and I think the order should always be key (`out`) then dependencies (`x`) so this one seems fine, but there should probably be a unit test that enforces that order as well.

zzhengnan replied:
Good to know that for this one instance it's safe to flip the order in the dict. Any suggestions for how to handle the same situation in the other PRs and the (I want to say, close to 100) places where we aren't yet sure whether the key ordering in the output dict is deterministic? Is there a simpler way than looking at the source code for all the relevant functions?

jsignell replied:
I guess you could run the tests a few times and see if they fail. But probably best to just leave potentially unordered dicts skipped for now. I opened an issue to get a sense of whether dask should care about dict ordering https://github.com/dask/dask/issues/7911

jsignell replied:
Did you mean to unskip the lines below as suggested?

zzhengnan replied:
Ahh I was too caught up in the broader discussion and forgot about the actual code.

Fixed. Thanks for catching.

jsignell commented on dask/optimization.py:
```diff
@@ -236,7 +236,7 @@ def inline(dsk, keys=None, inline_constants=True, dependencies=None):
 
-    >>> d = {'x': 1, 'y': (inc, 'x'), 'z': (add, 'x', 'y')} # doctest: +SKIP
+    >>> d = {'x': 1, 'y': (inc, 'x'), 'z': (add, 'x', 'y')}
     >>> inline(d)       # doctest: +SKIP
```
You could unskip this using a similar pattern to the one I described in `cull`

zzhengnan replied:
Done.

jsignell reviewed (commented):
Thank you for opening this  -and sorry it took a while for me to look at. I have a few comments that probably apply to the other PRs as well. I'll just review this one for now so that we can keep the conversation in one spot.

zzhengnan commented on dask/local.py:
```diff
@@ -61,12 +61,16 @@
 {'cache': {'x': 1, 'y': 2},
  'dependencies': {'w': {'z', 'y'}, 'x': set(), 'y': set(), 'z': {'x'}},
- 'dependents': {'w': set(), 'x': {'z'}, 'y': {'w'}, 'z': {'w'}},
+ 'dependents': defaultdict(None,
+                           {'w': set(),
+                            'x': {'z'},
+                            'y': {'w'},
+                            'z': {'w'}}),
```
I had to spread these out across multiple lines to match `pprint`s formatting. Tried the following thinking it would ignore white spaces to no avail. If there's better way, let me know.
https://docs.python.org/3/library/doctest.html#doctest.NORMALIZE_WHITESPACE

```python
>>> pprint.pprint(start_state_from_dask(dsk))  # doctest: +NORMALIZE_WHITESPACE
```

zzhengnan replied:
Done. Reverted to the original skipping behavior.

zzhengnan commented on dask/local.py:
```diff
@@ -143,12 +147,16 @@ def start_state_from_dask(dsk, cache=None, sortkey=None):
-    >>> dsk = {'x': 1, 'y': 2, 'z': (inc, 'x'), 'w': (add, 'z', 'y')}  # doctest: +SKIP
-    >>> from pprint import pprint  # doctest: +SKIP
-    >>> pprint(start_state_from_dask(dsk))  # doctest: +SKIP
+    >>> dsk = {'x': 1, 'y': 2, 'z': (inc, 'x'), 'w': (add, 'z', 'y')}
```
This seems to be the exact same example as the one shown in the module docstring. Is that something we don't generally mind?

zzhengnan reviewed (commented):
@jsignell Thanks for the review and great suggestions! I agree it's best to limit the conversation in one spot as opposed to spread out across all 6 PRs.

As you'll see from my follow-up comments, I have a few more clarifying questions. Once we get those figured out and this PR merged, I can proceed to making the same set of changes in the remaining PRs.

zzhengnan commented on dask/delayed.py:
```diff
@@ -141,6 +141,7 @@ def to_task_dask(expr):
 
     Examples
     --------
+    # TODO: Fix this doctest
```
Don't quite know how to fix this one.
```python
142     Examples
143     --------
        # TODO: Fix this doctest
144     >>> import dask
145     >>> a = delayed(1, 'a')
146     >>> b = delayed(2, 'b')
147
148     >>> task, dask = to_task_dask([a, b, 3])    
UNEXPECTED EXCEPTION: UserWarning('The dask.delayed.to_dask_dask function has been Deprecated in favor of unpack_collections')
Traceback (most recent call last):
  File "/Users/zhengnan/miniforge3/envs/dask-dev/lib/python3.8/doctest.py", line 1336, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest dask.delayed.to_task_dask[3]>", line 1, in <module>
  File "/Users/zhengnan/Desktop/open-source/dask/dask/delayed.py", line 160, in to_task_dask
    warnings.warn(
UserWarning: The dask.delayed.to_dask_dask function has been Deprecated in favor of unpack_collections
```

jsignell replied:
This one can probably stay skipped since the method is deprecated.

zzhengnan commented on dask/highlevelgraph.py:
```diff
@@ -124,6 +124,7 @@ def cull(
 
         Examples
         --------
+        TODO: Fix this doctest
         >>> d = Layer({'x': 1, 'y': (inc, 'x'), 'out': (add, 'x', 10)})  # doctest: +SKIP
```
Don't quite know how to fix this one.

```python
125         Examples
126         --------
127         TODO: Fix this doctest
128         >>> d = Layer({'x': 1, 'y': (inc, 'x'), 'out': (add, 'x', 10)})
UNEXPECTED EXCEPTION: TypeError("Can't instantiate abstract class Layer with abstract methods __getitem__, __iter__, __len__, get_output_keys, is_materialized")
Traceback (most recent call last):
  File "/Users/zhengnan/miniforge3/envs/dask-dev/lib/python3.8/doctest.py", line 1336, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest dask.highlevelgraph.Layer.cull[0]>", line 1, in <module>
TypeError: Can't instantiate abstract class Layer with abstract methods __getitem__, __iter__, __len__, get_output_keys, is_materialized
```

jsignell replied:
This seems to work for me:

```python
        >>> d = MaterializedLayer({'x': 1, 'y': (inc, 'x'), 'out': (add, 'x', 10)})
        >>> layer, deps = d.cull({'out'}, d.keys())
        >>> deps
        {'out': {'x'}, 'x': set()})
```

zzhengnan commented:
@jsignell I believe I addressed all your remaining comments. The only thing left is https://github.com/dask/dask/pull/7874#discussion_r670691412, which I don't think is a blocker for merging this PR.

Let me know if you have other suggestions/concerns. If not, I'll move on to #7873. Thanks!

jsignell commented:
Thanks for working on this @zzhengnan! Failures are unrelated see #7907

jsignell commented:
Thank you for working on this @zzhengnan!!

zzhengnan commented:
@jsignell Thanks for the guidance and all the great suggestions! 

I'll start incorporating what we discussed in this PR into #7873. Should I tag you once that PR is ready for review, or is another developer going to pick up the review?

jsignell commented:
You can tag me :)
